homeassistant/components/zha/core/gateway.py

```diff
@@ -89,26 +113,56 @@ def get_entities_for_ieee(self, ieee_str):
             return self._device_registry[ieee]
         return []
 
+    def get_entity(self, entity_id):
+        """Return entity for given entity_id if found."""
+        for entity in itertools.chain.from_iterable(
+                self.device_registry.values()):
+            if entity_id == entity.entity_id:
+                return entity
+
     @property
-    def device_registry(self) -> str:
+    def devices(self):
         """Return devices."""
+        return self._devices
+
+    @property
+    def device_registry(self):
+        """Return entities by ieee."""
         return self._device_registry
 
-    async def async_device_initialized(self, device, join):
+    def register_entity(self, ieee, entity_obj):
+        """Record the creation of a hass entity associated with ieee."""
+        self._device_registry[ieee].append(entity_obj)
+
+    async def _get_or_create_device(self, zigpy_device):
+        """Get or create a ZHA device."""
+        zha_device = self._devices.get(zigpy_device.ieee, None)
+        if zha_device is None:
+            zha_device = ZHADevice(self._hass, zigpy_device, self)
+            self._devices[zigpy_device.ieee] = zha_device
+        return zha_device
+
+    async def accept_zigbee_messages(self, _service_or_event):
+        """Allow devices to accept zigbee messages."""
+        for device in self.devices.values():
+            await device.async_accept_messages()
+
+    async def async_device_initialized(self, device, is_new_join):
         """Handle device joined and basic information discovered (async)."""
         import zigpy.profiles
 
-        device_manufacturer = device_model = None
-
+        zha_device = await self._get_or_create_device(device)
+        discovery_infos = []
         for endpoint_id, endpoint in device.endpoints.items():
             if endpoint_id == 0:  # ZDO
+                await _create_cluster_listener(
```

Comment:
not sure it is worth the effort. Many devices only have 2 endpoints and the most I have seen is 4. I can explore it if you really think it is worth it.

Reply:
any chance this causes any sort of deadlock like issues? If we have tasks that just keep fracturing out into scatter/gather?

Reply:
As long as one type of tasks can be executed out of order, we can still be sure that other types of tasks are not executed before, since we await the tasks before we continue.

I can't say anything specific since I don't know the device api.

Reply:
By principle we shouldn't await in loops if we don't need task order.